waku/v2/protocol/common_service.go

```diff
@@ -45,29 +49,44 @@ var ErrAlreadyStarted = errors.New("already started")
 var ErrNotStarted = errors.New("not started")
 
 // mutex protected stop function
-func (sp *CommonService) Stop(fn func()) {
+func (sp *CommonService[T]) Stop(fn func()) {
 	sp.Lock()
 	defer sp.Unlock()
-	if !sp.started {
+	if !sp.started.Load() {
 		return
 	}
 	sp.cancel()
 	fn()
 	sp.wg.Wait()
-	sp.started = false
+	close(sp.channel)
+	sp.started.Store(false)
 }
 
 // This is not a mutex protected function, it is up to the caller to use it in a mutex protected context
-func (sp *CommonService) ErrOnNotRunning() error {
-	if !sp.started {
+func (sp *CommonService[T]) ErrOnNotRunning() error {
+	if !sp.started.Load() {
 		return ErrNotStarted
 	}
 	return nil
 }
 
-func (sp *CommonService) Context() context.Context {
+func (sp *CommonService[T]) Context() context.Context {
 	return sp.ctx
 }
-func (sp *CommonService) WaitGroup() *sync.WaitGroup {
+func (sp *CommonService[T]) WaitGroup() *sync.WaitGroup {
 	return &sp.wg
 }
+func (sp *CommonService[T]) GetListeningChan() <-chan T {
+	return sp.channel
+}
+func (sp *CommonService[T]) PushToChan(data T) bool {
+	if !sp.started.Load() {
```

Comment:
This check would not be sufficient. 
Let's say stop was called and started is still not set to false, then this check will go through and stop will also continue. This may cause race condition. You will have to use the readMutex to protect this.